lib/services/local/perf_test.go

```diff
@@ -0,0 +1,137 @@
+/*
+Copyright 2020 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package local
+
+import (
+	"context"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"testing"
+
+	"github.com/gravitational/teleport/lib/backend"
+	"github.com/gravitational/teleport/lib/backend/lite"
+	"github.com/gravitational/teleport/lib/backend/memory"
+	"github.com/gravitational/teleport/lib/defaults"
+	"github.com/gravitational/teleport/lib/services"
+
+	"github.com/stretchr/testify/assert"
+)
+
+// BenchmarkGetNodes verifies the performance of the GetNodes operation
+// on local (sqlite) databases (as used by the cache system).
+func BenchmarkGetNodes(b *testing.B) {
+
+	type testCase struct {
+		validation, memory bool
+		nodes              int
+	}
+
+	var tts []testCase
+
+	for _, validation := range []bool{true, false} {
+		for _, memory := range []bool{true, false} {
+			for _, nodes := range []int{100, 1000, 10000} {
+				tts = append(tts, testCase{
+					validation: validation,
+					memory:     memory,
+					nodes:      nodes,
+				})
+			}
+		}
+	}
+
+	for _, tt := range tts {
+		// create a descriptive name for the sub-benchmark.
+		name := fmt.Sprintf("tt(validation=%v,memory=%v,nodes=%d)", tt.validation, tt.memory, tt.nodes)
+
+		// set up marshal options
+		var opts []services.MarshalOption
+		if !tt.validation {
+			opts = append(opts, services.SkipValidation())
+		}
+
+		// configure the backend instance
+		var bk backend.Backend
+		var err error
+		if tt.memory {
+			bk, err = memory.New(memory.Config{})
+			assert.NoError(b, err)
+		} else {
+			dir, err := ioutil.TempDir("", "teleport")
+			assert.NoError(b, err)
+			defer os.RemoveAll(dir)
+
+			bk, err = lite.NewWithConfig(context.TODO(), lite.Config{
+				Path: dir,
+			})
+			assert.NoError(b, err)
+		}
+		defer bk.Close()
+
+		svc := NewPresenceService(bk)
+
+		// seed the test nodes
+		insertNodes(b, svc, tt.nodes)
```

Comment:
optional: you can put all of this inside `b.Run` so that defers run sooner and use `b.StopTimer/b.StartTimer` to avoid measuring the setup time.
that way the tests don't affect each other via GC pressure or any other background goroutines (from unclosed backends) consuming CPU.